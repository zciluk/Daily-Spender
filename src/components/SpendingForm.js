import React, { Component } from "react";
import { Box, FormField, Button, Form } from "grommet";
import { Add } from "grommet-icons";
import propTypes from 'prop-types';
class SpendingForm extends Component {
  onSubmit = value => {
    this.props.addNewSpending(value);
  };

  render() {
    return (
      <Box
        direction="column"
        align="center"
        pad={{ left: "medium", right: "small", vertical: "small" }}
        style={{ zIndex: "1" }}
      >
        <Form onSubmit={({ value }) => this.onSubmit(value)}>
          <FormField name="name" placeholder="Name" required />
          <FormField
            name="value"
            placeholder="Value"
            required
            validate={{
              regexp: /^[0-9]*$/,
              message: "only numbers are acceptable"
            }}
          />
          <Button
            type="submit"
            icon={<Add />}
            margin="small"
            alignSelf="center"
            label="Add spending"
            name="submit"
            onClick={() => {}}
          />
        </Form>
      </Box>
    );
  }
}
SpendingForm.propTypes = {
  addNewSpending: propTypes.func
};


export default SpendingForm;
