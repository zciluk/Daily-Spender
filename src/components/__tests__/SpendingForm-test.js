import React from 'react';
import SpendingForm from '../SpendingForm.js';
import { mount} from 'enzyme';
import renderer from 'react-test-renderer';
import sinon from 'sinon';

// ****
// SNAPSHOT UI TESTS - JEST
// ****
test('should render initial layout', () => {
    const component = renderer.create(
        <SpendingForm/>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
    // snapshot checks if placeholders are properly displayed, input fields are cleared and button is displayed correctly
});


// should display errors when click Add spending 
test('should display errors when click Add spending with empty fields', () => { 
    const component = renderer.create(
        <SpendingForm/>
    );
    const fakeEvent = { preventDefault: () => console.log() };
    component.root.findByType('form').props.onSubmit(fakeEvent);

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
    // snapshots are awesome! this one is checking whether Required spans are properly displayed

});
//

// ****
// OTHER TESTS - ENZYME
// ****
test('should display error when letters are input into Value fields', () => { 
    const component = mount(
        <SpendingForm />
    );
    component.find('input[name="value"]').simulate('change', { target: { value: 'test' } });
    component.find('form').simulate('submit', { preventDefault () {} });
    expect(component.find('span').at(1).text()).toEqual("only numbers are acceptable");
});

test('should return proper onSubmit function when fields are filled properly', () => { 
    const callback = sinon.spy();
    const component = mount(
        <SpendingForm addNewSpending={callback} />
    );
    var expectedResult = {
        name: "test",
        value: "123"
    };
    
    component.find('input[name="name"]').simulate('change', { target: { value: 'test' } });
    component.find('input[name="value"]').simulate('change', { target: { value: '123' } });
    component.find('button[type="submit"]').simulate('click');
    component.find('form').simulate('submit', { preventDefault () {} });
    sinon.assert.calledWith(callback, expectedResult);
});