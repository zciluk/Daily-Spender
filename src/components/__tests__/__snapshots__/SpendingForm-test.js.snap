// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display errors when click Add spending with empty fields 1`] = `
<div
  className="StyledBox-sc-13pk1d4-0 eWmAkr"
  style={
    Object {
      "zIndex": "1",
    }
  }
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="StyledBox-sc-13pk1d4-0 CAsMy"
    >
      <div
        className="StyledBox-sc-13pk1d4-0 fVtecF"
      >
        <div
          className="StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 yIdXy"
        >
          <input
            autoComplete="off"
            className="StyledTextInput-sc-1x30a0s-0 cuRQxH"
            name="name"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            placeholder="Name"
            value=""
          />
        </div>
      </div>
      <div
        className="StyledBox-sc-13pk1d4-0 fsgjrx"
      >
        <span
          className="StyledText-sc-1sadyjn-0 hOyydb"
        >
          required
        </span>
      </div>
    </div>
    <div
      className="StyledBox-sc-13pk1d4-0 CAsMy"
    >
      <div
        className="StyledBox-sc-13pk1d4-0 fVtecF"
      >
        <div
          className="StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 yIdXy"
        >
          <input
            autoComplete="off"
            className="StyledTextInput-sc-1x30a0s-0 cuRQxH"
            name="value"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            placeholder="Value"
            value=""
          />
        </div>
      </div>
      <div
        className="StyledBox-sc-13pk1d4-0 fsgjrx"
      >
        <span
          className="StyledText-sc-1sadyjn-0 hOyydb"
        >
          required
        </span>
      </div>
    </div>
    <button
      className="StyledButton-sc-323bzc-0 fjELoD"
      name="submit"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      type="submit"
    >
      <div
        className="StyledBox-sc-13pk1d4-0 hHwlHq"
      >
        <svg
          aria-label="Add"
          className="StyledIcon-ofa7kd-0 iOkQrb"
          viewBox="0 0 24 24"
        >
          <path
            d="M12,22 L12,2 M2,12 L22,12"
            fill="none"
            stroke="#000"
            strokeWidth="2"
          />
        </svg>
        <div
          className="StyledBox__StyledBoxGap-sc-13pk1d4-1 iChEkS"
        />
        Add spending
      </div>
    </button>
  </form>
</div>
`;

exports[`should render initial layout 1`] = `
<div
  className="StyledBox-sc-13pk1d4-0 eWmAkr"
  style={
    Object {
      "zIndex": "1",
    }
  }
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="StyledBox-sc-13pk1d4-0 CAsMy"
    >
      <div
        className="StyledBox-sc-13pk1d4-0 gZSidW"
      >
        <div
          className="StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 yIdXy"
        >
          <input
            autoComplete="off"
            className="StyledTextInput-sc-1x30a0s-0 cuRQxH"
            name="name"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            placeholder="Name"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      className="StyledBox-sc-13pk1d4-0 CAsMy"
    >
      <div
        className="StyledBox-sc-13pk1d4-0 gZSidW"
      >
        <div
          className="StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 yIdXy"
        >
          <input
            autoComplete="off"
            className="StyledTextInput-sc-1x30a0s-0 cuRQxH"
            name="value"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            placeholder="Value"
            value=""
          />
        </div>
      </div>
    </div>
    <button
      className="StyledButton-sc-323bzc-0 fjELoD"
      name="submit"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      type="submit"
    >
      <div
        className="StyledBox-sc-13pk1d4-0 hHwlHq"
      >
        <svg
          aria-label="Add"
          className="StyledIcon-ofa7kd-0 iOkQrb"
          viewBox="0 0 24 24"
        >
          <path
            d="M12,22 L12,2 M2,12 L22,12"
            fill="none"
            stroke="#000"
            strokeWidth="2"
          />
        </svg>
        <div
          className="StyledBox__StyledBoxGap-sc-13pk1d4-1 iChEkS"
        />
        Add spending
      </div>
    </button>
  </form>
</div>
`;
